### PR TITLE
tests, infra_test, Fix flaky test 4134

### PR DESCRIPTION
**Pull request by oshoval**
**What this PR does / why we need it**:

When test 4134 runs on check-tests-for-flakes
it might fail on a panic, because when it sampled
for nodes, there was a hiccup in the API.

Add a retry in case needed by using Eventually,
in order to eliminate the flakiness.

Removes the need to copy the nodes to
a helper array.

Improve the AfterEach to check if the selectedNode
is not nil before dereferencing it.

The loop search for a pod on a non master node,
once it is found, no need to continue iteration.

Fixes: https://github.com/kubevirt/kubevirt/issues/4158

Signed-off-by: Or Shoval <oshoval@redhat.com>

**Special notes for your reviewer**:

```release-note
None
```

### CODE REVIEW

File: tests/infra_test.go [hunk 1]
```diff
@@ -221,19 +221,24 @@ var _ = Describe("Infrastructure", func() {
...
 
 			AfterEach(func() {
-				By("removing the taint from the tainted node")
-				var taints []k8sv1.Taint
-
-				for _, taint := range selectedNode.Spec.Taints {
-					if taint.Key != "CriticalAddonsOnly" {
-						taints = append(taints, taint)
+				if selectedNode != nil {
+					defer func() {
```
Comment (RamLavi) [inline tests/infra_test.go:225]: when does the defer triggered in tests? is it after each test (i.e. at the end of every `It`)?

Comment (RamLavi) [reply tests/infra_test.go:225]: I think i know the answer - is it after the context is closed. in this case it'll be in the end of the `AfterEach`, right?

Comment (oshoval) [reply tests/infra_test.go:225]: yes, its in AfterEach , just in case selectedNode is not nil (this what the deferred func does, just nil it, in case of exceptions)

Comment (RamLavi) [reply tests/infra_test.go:225]: [yep](https://tour.golang.org/flowcontrol/12)

Comment (oshoval) [reply tests/infra_test.go:225]: i might need to remove it, in case it doesnt change the ptr as in C (passed by val versus passed by ref)

Comment (oshoval) [reply tests/infra_test.go]: removed, we also already have
```
			BeforeEach(func() {
				selectedNode = nil
			})
```

File: tests/infra_test.go [hunk 2]
```diff
@@ -252,24 +253,26 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
+				Eventually(func() *k8sv1.Node {
+					schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
+					for _, node := range schedulableNodesList.Items {
+						if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
+							continue
+						}
+						for _, pod := range pods.Items {
+							if pod.Spec.NodeName == node.Name {
+								selectedNode = node.DeepCopy()
+								return selectedNode
```
Comment (RamLavi) [inline tests/infra_test.go:269]: in the code prior to your change, selectedNode will be given to the node that hold the last pod in the loop (since we're iterating over the pods and set selectedNode each time).
in your changed code, selectedNode will be given to the node that hold the first pod in the loop.
This could have the consistent outcome only if there is 1 pod in the list..

can you explain why this change is ok?

Comment (oshoval) [reply tests/infra_test.go:269]: because we just need a node which isn't the master node, the previous code just continued iterating for no reason at all.

**Comment by oshoval** [comment]
Thanks Ram
/retest

File: tests/infra_test.go [hunk 3]
```diff
@@ -252,24 +253,26 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
 				// tolerate this taint because it is meant to be used on compute nodes only. If we set this taint
 				// on a master node, we risk in breaking the test cluster.
-				for _, pod := range pods.Items {
-					node := schedulableNodes[pod.Spec.NodeName]
-					if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
-						continue
+				Eventually(func() *k8sv1.Node {
+					schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
+					for _, node := range schedulableNodesList.Items {
+						if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
```
Comment (qinqon) [inline tests/infra_test.go:263]: What about filtering by `schedulable` and `node-role` specified two label selector at find ? for that we will need some refactoring.

File: tests/infra_test.go [hunk 4]
```diff
@@ -252,24 +253,26 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
 				// master nodes should never have the CriticalAddonsOnly taint because core components might not
 				// tolerate this taint because it is meant to be used on compute nodes only. If we set this taint
 				// on a master node, we risk in breaking the test cluster.
-				for _, pod := range pods.Items {
-					node := schedulableNodes[pod.Spec.NodeName]
-					if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
-						continue
+				Eventually(func() *k8sv1.Node {
+					schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
+					for _, node := range schedulableNodesList.Items {
```
Comment (qinqon) [inline tests/infra_test.go:262]: Can you name `node` as `schedulableNode` so the loop code is more clear.

File: tests/infra_test.go [hunk 5]
```diff
@@ -252,24 +253,26 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
+						if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
+							continue
+						}
+						for _, pod := range pods.Items {
```
Comment (qinqon) [inline tests/infra_test.go:266]: pods here are kubevirtPods so let's rename this `for _, kubevirtPod := range kubevirtPods.Items`

this way we will have `if kubevirtPod.Spec.NodeName == scheduableNode...` that's better to understand the loop.

**Review by qinqon — CHANGES_REQUESTED** [review]
Looks good just naming and a proposal.

File: tests/infra_test.go (hunk 3 above)
Comment (oshoval) [reply tests/infra_test.go:263]: as we spoke, this is something we can check and improve on another cycle, but not the purpose of this PR, nor changed but just reordered a bit.
also its pretty readable right now.

**Comment by oshoval** [comment]
Rebased only.

File: tests/infra_test.go (hunk 4 above)
Comment (oshoval) [reply tests/infra_test.go]: Done

File: tests/infra_test.go [hunk 6]
```diff
@@ -226,50 +226,53 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
-
-				for _, taint := range selectedNode.Spec.Taints {
-					if taint.Key != "CriticalAddonsOnly" {
-						taints = append(taints, taint)
+				if selectedNode != nil {
+					By("removing the taint from the tainted node")
+					var taints []k8sv1.Taint
+					for _, taint := range selectedNode.Spec.Taints {
+						if taint.Key != "CriticalAddonsOnly" {
+							taints = append(taints, taint)
```
Comment (oshoval) [inline tests/infra_test.go:234]: not related to this PR, but is it just me or no one use the filtered taints ?

File: tests/infra_test.go [hunk 7]
```diff
@@ -226,50 +226,53 @@ var _ = Describe("[Serial]Infrastructure", func() {
...
 				By("selecting one compute only node that runs kubevirt components")
 				// master nodes should never have the CriticalAddonsOnly taint because core components might not
 				// tolerate this taint because it is meant to be used on compute nodes only. If we set this taint
 				// on a master node, we risk in breaking the test cluster.
-				for _, pod := range pods.Items {
-					node := schedulableNodes[pod.Spec.NodeName]
-					if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
-						continue
+				Eventually(func() *k8sv1.Node {
+					schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
```
Comment (AlonaKaplan) [inline tests/infra_test.go:261]: How can we get to the situation we have pods but no schedulable nodes?

Comment (oshoval) [reply tests/infra_test.go:261]: see here that it happened,
https://github.com/kubevirt/kubevirt/issues/4158
when it runs in check-tests-for-flakes, its the first test that runs on a fresh cluster, and then the cluster seems sometimes to have a little hiccup after it has the pod list, for a short time (it happens on fresh clusters sometimes from what i saw in the past, and it looks like it what happens here).

Comment (AlonaKaplan) [reply tests/infra_test.go:261]: So it seems that we have a bigger problem. We might start running the tests on a not ready cluster.
We cannot add eventually's to each one of the tests waiting for the cluster to be ready.

I think we should understand the root cause of the problem and fix it. Fixing this specific test is confusing and doesn't solve the problem for the other tests that might have the same issue.

Comment (oshoval) [reply tests/infra_test.go:261]: We didn't see it hurt other tests explicitly.
This test is sensitive for it and it cause flakes on the `check-lane-for-flakes` which runs only the files that are changed in the PR, and then it can be the first test, so i think its fine and good to have this test stable, in case we need we can do a larger scale fix on other PR.

Comment (AlonaKaplan) [reply tests/infra_test.go:261]: The PR can keep the other refactories it has. But I think the use of the `Eventually` is wrong and may hide other problems.
The test shouldn't assume the cluster may have "hiccups". We shouldn't start adding `Eventually`s to our tests to cover those "hiccups".
Can you please explain me what is so special in this test that makes it "especially sensitive"? Should we now fix all the tests the are using `tests.GetAllSchedulableNodes`?

I believe that the issue that you found it bigger than this specific test. We are running the tests too early. And that's what I believe should be fixed.

Comment (oshoval) [reply tests/infra_test.go:261]: > The PR can keep the other refactories it has. But I think the use of the `Eventually` is wrong and may hide other problems.
> The test shouldn't assume the cluster may have "hiccups". We shouldn't start adding `Eventually`s to our tests to cover those "hiccups".
> Can you please explain me what is so special in this test that makes it "especially sensitive"? Should we now fix all the tests the are using `tests.GetAllSchedulableNodes`?

Its special that when it runs as the first test its more sensitive than others (from what we actually see, not about what we didn't see yet), as it has a flow where it samples the nodes scheduable flag, which other tests doesnt do directly, but implicit on some stages in their life time maybe.

> 
> I believe that the issue that you found it bigger than this specific test. We are running the tests too early. And that's what I believe should be fixed.

It was said implicit in the above text "This test is especially sensitive for it and it cause flakes on the check-lane-for-flakes which runs only the files that are changed in the PR, and then it can be the first test".
But i think we should first see it happens and affect other tests and then act, at least to some degree, or if we simulate it, we also need to prioritize it, i wont suggest to put it highest priority before we dive into it more and see it harms (it might harm here and there already, maybe the compile time for example of all the tests which is longer that just compiling this test, make the transient go
).
I honestly don't think we should block this PR, nor suggested to add 'Eventually' on other PRs, here its suitable, and natural as part of the other changes, in order to reduce noise. 
We already learned from it what we should and can act ortogonal.

Comment (AlonaKaplan) [reply tests/infra_test.go:261]: Sorry, but I'm totally against putting an `Eventually` to cover an error we don't fully understand just to reduce the noise.
Using `Eventually` makes sense when we are waiting for a result after an async update.
Here it is not the case. We are waiting for schedulable nodes on a cluster with running pods!
All our tests assume the cluster is ready, I truly believe it is wrong to soften this assumption on some tests just to reduce noise.

Comment (oshoval) [reply tests/infra_test.go:261]: I understand, i just think that since we dont see the error affecting other tests, but just this one, and just on the `check-for-flakes` lane, and not even on flake-finder, i found it safe, else i would think the same.
